libr/core/disasm.c

```diff
@@ -6583,3 +6583,186 @@ R_API int r_core_disasm_pdi(RCore *core, int nb_opcodes, int nb_bytes, int fmt)
+	ut64 block_start = r_reg_get_value (reg, pc);
+	ut8 *buf = malloc (buf_sz);
+	int i;
+	for (i = 0; i < nb_opcodes; i++) {
+		const ut64 op_addr = r_reg_get_value (reg, pc);
+		if (block_sz + read_len > buf_sz) {
+			buf_sz *= 2;
+			ut8 *tmp = realloc (buf, buf_sz);
+			if (!tmp) {
+				break;
+			}
+			buf = tmp;
+		}
+		if (!r_io_read_at_mapped (core->io, op_addr, buf + block_sz, read_len)) {
+			break;
+		}
+		RAnalOp op;
+		int ret = r_anal_op (core->anal, &op, op_addr, buf + block_sz, read_len, R_ANAL_OP_MASK_ESIL);
+		const bool invalid_instr = ret < 1 || op.size < 1;
+		bool end_of_block = false;
+		switch (op.type & R_ANAL_OP_TYPE_MASK & ~R_ANAL_OP_HINT_MASK) {
+		case R_ANAL_OP_TYPE_JMP:
+		case R_ANAL_OP_TYPE_UJMP:
+		case R_ANAL_OP_TYPE_CALL:
+		case R_ANAL_OP_TYPE_UCALL:
+		case R_ANAL_OP_TYPE_RET:
+			end_of_block = true;
+			if (hide_branches) {
+				i--;
+			} else {
+				block_instr++;
+				block_sz += op.size;
+			}
+			break;
+		default:
+			if (i + 1 >= nb_opcodes || invalid_instr) {
+				end_of_block = true;
+			}
+			block_instr++;
+			block_sz += op.size;
+			break;
+		}
+		if (end_of_block) {
+		print_delay_slots:
+			if (block_instr) {
+				switch (mode) {
+				case R_MODE_JSON:
+					r_core_print_disasm_json (core, block_start, buf, block_sz, block_instr, pj);
+					break;
+				case R_MODE_SIMPLE:
+					r_core_seek (core, block_start, true);
+					r_core_disasm_pdi (core, block_instr, block_sz, 0);
+					break;
+				case R_MODE_SIMPLEST:
+					r_core_seek (core, block_start, true);
+					r_core_print_disasm_instructions (core, block_sz, block_instr);
+					break;
+				default:
+					r_core_print_disasm (core->print, core, block_start, buf, block_sz, block_instr, 0, block_sz, false, NULL, NULL);
+					break;
+				}
+			}
+			if (op.delay) {
+				block_instr = R_MIN (op.delay, nb_opcodes - (i + 1));
+				if (!block_instr) {
+					break;
+				}
+				block_sz = block_instr * read_len;
+				if (block_sz > buf_sz) {
+					buf_sz *= 2;
+					ut8 *tmp = realloc (buf, buf_sz);
+					if (!tmp) {
+						break;
+					}
+					buf = tmp;
+				}
+				if (!r_io_read_at_mapped (core->io, op.addr + op.size, buf, block_sz)) {
+					break;
+				}
+				block_start = op.addr + op.size;
+				i += block_instr;
+				op.delay = 0;
+				goto print_delay_slots;
+			}
+			block_sz = 0;
+			block_instr = 0;
+		}
+		if (invalid_instr) {
+			break;
+		}
+		r_anal_esil_set_pc (core->anal->esil, op_addr);
+		r_reg_set_value (reg, pc, op_addr + op.size);
+		const char *e = r_strbuf_get (&op.esil);
+		if (R_STR_ISNOTEMPTY (e)) {
+			r_anal_esil_parse (esil, e);
+		}
+		r_anal_op_fini (&op);
+
+		if (end_of_block) {
+			block_start = r_reg_get_value (reg, pc);
+			r_core_seek_arch_bits (core, block_start);
+		}
+	}
+	if (mode == R_MODE_JSON) {
+		pj_end (pj);
+		r_cons_print (pj_string (pj));
+		pj_free (pj);
+	}
+	free (buf);
+	r_reg_arena_pop (reg);
+	int len = r_list_length (ocache);
+	if (r_list_length (core->io->cache) > len) {
```

Comment:
uhm, this seems like it should be implemented in r_io_cache or r_cache as apis intead of this inlined code